### ISSUE TITLE
Consider some mechanism to get limited-access to syntax tree green-nodes.

### ISSUE THREAD

**Issue by CyrusNajmabadi** [dotnet/roslyn#63172]
## Background and Motivation

The Roslyn syntax model is incredibly powerful and capable.  But it comes with a caveat.  It is extremely expensive in terms of memory and allocations.  This is due to a split-object model that produces underlying 'green nodes' from a parser, which are then wrapped on-demand into 'red nodes' when the tree is walked.  The primary differences between the nodes are the 'green nodes' do not have a parent, nor do they have a position.  This allows them to trivially be reused in incremental scenarios (even if they need to move around in the tree), as well allowing for the nodes to appear multiple times in the same tree, and even across trees.  This makes the green-tree (Actually a green-directed-acyclic-graph) be very small and lightweight, though far less easy to use.   Conversely, the red-tree is much more flexible, allowing clients to get to any place in the tree from any other place, and to know precisely where nodes sit in the original file.

There are times though when features do not need the features of the red-tree and would benefit greatly from not taking the hit of creating red-nodes.  A good example of this is 'indexing'.  e.g. any client that wants to walk the tree to determine certain interesting attributes about it.  We already use the green-tree effectively in this way for the new incremental-generator api `ForAttributeWithMetadataName`.  This API incrementally builds an index which knows which files have attributes in it, versus which do not.  This is dramatically sped up being able to walk the green tree instead of the red tree (GBs of savings of allocations in a roslyn-sized solution as edits are coming in).

Performance testing shows this is also significant for the IDE to produce our indices (For example, the indices that power NavTo and FindRefs).  These indices are built just by walking down the syntax tree, collecting useful information (like what identifiers were in the file).  Today, producing this index for roslyn on workstation-gc is heavily impacted by the allocation cost of the red-tree.  For comparison, to produce all the indices for Roslyn.sln:

1. Using Red-trees take 35 seconds.
2. Using Green-trees takes 19 seconds.  

This is a 45% decrease in time, and a *lot* less impact on the memory subsystem and GC.

For context, this is for 450MB of source, which can easily explode in teh amount of memory allocated given how heavyweight trees are.

## Proposed API

The goal here is to bring to API review to discuss potential options.  This is likely very contentious, so we would need to discuss in depth what feels ok to expose and how to do it.  For some context, the IDE just needs to walk down the tree, checking kinds, and getting things like the .ValueText of some SyntaxTokens.  In this world, we would only need a tiny exposed surface area to make this possible.

However, other contexts, the IDE likely needs to know the actual 'green node type' of the node so we can pull out relevant sub-bits in a strongly typed fashion.  For example, getting a green-ClassDeclarationSyntax node so we can grab the .Identifier, look at the .TypeParameters, etc. etc. etc.  Exposing this would *greatly* increase the surface area exposed.

## Usage Examples

``` C#
// some lines of code here
```

**Comment by RikkiGibson** [comment]
> Today, producing this index for roslyn on workstation-gc is heavily impacted by the allocation cost of the red-tree. For comparison, to produce all the indices for Roslyn.sln:
> 
> 1. Using Red-trees take 35 seconds.
> 2. Using Green-trees takes 19 seconds.

How did you determine this?

**Comment by CyrusNajmabadi** [comment]
i simply wrote a benchmark that produced this.  You can see this in things like NavigateToBenchmarks.cs.  I then run the benchmark while running PerfView and i can then see the relative costs of things (which tells me where i should be investing in).  This is also how i did much of the recent work around Incrementl-Generator perf.

**Comment by RikkiGibson** [comment]
Makes sense. I wasn't sure if you had written a prototype which uses publicly-exposed green nodes and benchmarked that to compare.

**Comment by CyrusNajmabadi** [comment]
Oh yes. I also did that. I just made everything I needed public and rewrote the indexer to sit on that :-)

**Comment by CyrusNajmabadi** [comment]
Note: generally speaking I think it's a stretch to do this. If we really decide we need the perf, then we can consider this. But for now, is say it's probably not worth it.
